Update @wordpress and React packages [MAILPOET-5102]

costasovo opened the pull request:
## Description
This PR updates @wordpres/* and React packages. React update is included because some of the wordpress packages [have already updated to React 18](https://github.com/WordPress/gutenberg/pull/45235) and use the new functionality.

Note: After the update, there is the dev console for dev builds.
```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot
```
This is expected for now and will be addressed in a follow-up ticket. Switching to the new render API will turn on the automatic batching feature, and it may cause some issues.


## Code review notes

_N/A_

## QA notes

This changes a lot of JS libraries we build on, so potentially, this could cause issues in the whole plugin.
Five parts of the plugin are influenced the most because they heavily use these packages:

1. Automation - some visual changes are expected
2. Form Editor - some visual changes are expected
3. Plugin settings
4. Dynamic segments editor (the form for creating segments)
5. Settings panel of the coupon block - some small visual changes are expected

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/692 - related changes in the premium plugin
https://github.com/mailpoet/wordpress-images/pull/43 - update of the image for CI builds

## Linked tickets

[MAILPOET-5102]

## After-merge notes

This also updates the node version in the dev environment so ping devs to update the local node and rebuild the WordPress image in their dev env.


[MAILPOET-5102]: https://mailpoet.atlassian.net/browse/MAILPOET-5102?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Update WordPress JS dependencies and React

Most of the WordPress packages already upgraded to react 18 and use
new hooks so we need to upgrade also the react libs.
[MAILPOET-5102]

costasovo committed:
Fix Can't resolve 'process/browser'

[MAILPOET-5102]

costasovo committed:
Update Custom font selector toolbar in form editor

This commit fixes CSS which was preventing clicking the custom font select component
and suppresses a deprecation warning by adding __nextUnconstrainedWidth property.
[MAILPOET-5102]

costasovo committed:
Unify form editor styles panels labels with post editor labels

[MAILPOET-5102]

costasovo committed:
Fix radio buttons groups to display inline in form editor

[MAILPOET-5102]

costasovo committed:
Update classnames in automations editor

[MAILPOET-5102]

costasovo committed:
Update dev env node version to 19

After the updates some packages that we use in zip build don't work
with node17. This commit updates dev env to the most recent node version.
[MAILPOET-5102]

JanJakes committed:
Make @wordpress/data type definitions compatible with current version

[MAILPOET-5102]

JanJakes committed:
Use new @wordpress/data types in automations

[MAILPOET-5102]

JanJakes committed:
Use new @wordpress/data types in form editor

[MAILPOET-5102]

costasovo committed:
Use new @wordpress/data types in homepage

[MAILPOET-5102]

costasovo committed:
Use new @wordpress/data types in dynamic segments

[MAILPOET-5102]

costasovo committed:
Use new @wordpress/data types in form templates app

[MAILPOET-5102]

costasovo committed:
Fix ErrorBoundary type errors

[MAILPOET-5102]

costasovo committed:
Fix this.context is unknown error in send.tsx

[MAILPOET-5102]

costasovo committed:
Fix TS issues related to core/notice store

@types/wordpress__notice uses @types/wordpress_data but these are no
longer updated and combination with build types from the latest @wordpress/data
causes issues.
[MAILPOET-5102]

costasovo committed:
Fix Type errors related to block-editor store

[MAILPOET-5102]

costasovo committed:
Fix dependency cycles in form editor

[MAILPOET-5120]

costasovo committed:
Refactor settings store init to use new api

[MAILPOET-5102]

costasovo committed:
Fix rendering options in option-button

[MAILPOET-5102]

costasovo committed:
Fix types in form editor hooks

[MAILPOET-5102]

JanJakes committed:
Add typings for all @wordpress/data store selection variants

Both store name and store descriptor typings should work in all cases.

[MAILPOET-5102]

JanJakes committed:
Reintroduce store name as identifier in automation listing store

[MAILPOET-5102]

JanJakes committed:
Reintroduce store name as identifier in settings store

[MAILPOET-5102]

costasovo committed:
Reintroduce store name as identifier in form editor and templates stores

[MAILPOET-5102]

costasovo committed:
Reintroduce store name as identifier in dynamic segments store

[MAILPOET-5102]

costasovo committed:
Reintroduce store name as identifier in dynamic segments store

[MAILPOET-5102]

costasovo committed:
Reintroduce store name as identifier in automation editor store

[MAILPOET-5102]

JanJakes committed:
Add store types for @wordpress packages

[MAILPOET-5102]

JanJakes committed:
Fix types for @wordpress/notices

[MAILPOET-5102]

costasovo committed:
Fix type issues in dynamic segments

[MAILPOET-5102]

costasovo committed:
Fix unknown properties type issues

[MAILPOET-5102]

costasovo committed:
Fix rest of Typescript issues

[MAILPOET-5102]

costasovo committed:
Alter export of modules from dynamic segments for premium and extend types

[MAILPOET-5102]

costasovo committed:
Remove unused OmitFirstArgs and OmitFirstArg type helpers

[MAILPOET-5102]

JanJakes committed:
Add extensible interface for automation editor store

[MAILPOET-5102]

JanJakes committed:
Use relative imports that are resolvable for types in premium

[MAILPOET-5102]

JanJakes commented on mailpoet/assets/js/src/automation/components/option-button/index.tsx:
```diff
@@ -38,7 +38,7 @@ export function OptionButton({
       {slots.length > 0 &&
         slots.map(({ key, slot }) => (
-          <Fragment key={`slot-${key}`}>{slot}</Fragment>
+          <Fragment key={`slot-${key}`}>{slot()}</Fragment>
```
Oh, did this work before? I also wonder if there's a difference between `{slot()}` and `<Slot>`...

JanJakes commented on mailpoet/assets/js/src/automation/editor/components/header/document_actions.tsx:
```diff
@@ -48,9 +40,9 @@ function DocumentActions({ children }): JSX.Element {
+              placement: 'bottom-start',
+              anchor: titleRef.current,
             }}
-            position="bottom center"
```
More of a note to self to check if the dropdown positioning in the UI somehow changed.

JanJakes commented on mailpoet/.babelrc:
```diff
@@ -1,8 +1,8 @@
     ["@babel/preset-react", { "runtime": "automatic" }],
-    "@babel/preset-env"
+    "@babel/preset-env",
+    ["@babel/preset-typescript", { "allowDeclareFields": true }]
```
I wonder whether we should review the preset order in any other config (for any reason)...

JanJakes commented on mailpoet/assets/js/src/form_editor/components/sidebar/sidebar.tsx:
```diff
@@ -1,9 +1,21 @@
 import { PlacementSettingsSidebar } from './placement_settings_sidebar';
 import { store } from '../../store';
 
+// workaround for block editor store useSelect
```
Was this later removed (moved to `types/...` as per https://github.com/mailpoet/mailpoet/pull/4786/commits/5638e314cd633498efe6a247dcc6328775d77175)? Need to check.

JanJakes commented on mailpoet/assets/js/src/form_editor/components/sidebar/default_sidebar.tsx:
```diff
@@ -8,6 +12,14 @@ import { BlockSettings } from './block_settings.jsx';
 import { SidebarHeader } from './sidebar_header';
 import { store } from '../../store';
 
+// workaround for block editor store useSelect
```
Was this later removed (moved to `types/...` as per https://github.com/mailpoet/mailpoet/pull/4786/commits/5638e314cd633498efe6a247dcc6328775d77175)? Need to check.

JanJakes reviewed (commented):
Great job! 👏

I added a few small notes. I can verify these myself, given that you'll be AFK, @costasovo.
Thanks!

JanJakes committed:
Render automation name dropdown centered

[MAILPOET-5102]

JanJakes committed:
Remove forgotten workarounds

[MAILPOET-5102]

JanJakes committed:
Fix font size picker trying to process numbers as strings

This fixes the "fontSizeValue.startsWith is not a function" error. It seems that the font size picker
now always expects strings, while the values from SETTINGS_DEFAULTS could be numbers,
and we've been using numbers throughout our codebase as well.

This also enables the usage of any font size units, such as em and rem.

This may be related to: https://github.com/WordPress/gutenberg/issues/44857

[MAILPOET-5102]

JanJakes committed:
Simplify slot rendering in automations

[MAILPOET-5102]

Aschepikov commented:
Console error when I am pressing on add a condition

![image](https://user-images.githubusercontent.com/113185437/225321771-2d4e3868-dd29-445d-ab23-3e85de8a1ec7.png)

and some tiny issues 

1. In the automations incorrect tags list
![image](https://user-images.githubusercontent.com/113185437/225319011-bb76aad0-fef5-490c-9f4d-dfda329e3416.png)
On prod 
![image](https://user-images.githubusercontent.com/113185437/225319048-6784f928-75fb-4e96-a262-3dfe5e8dadc5.png)

2. Automations background is not blurred when an element is deleting
![image](https://user-images.githubusercontent.com/113185437/225319116-6becf8e0-86bf-401e-bcb3-4499e9347621.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225319152-c8581eed-718c-4021-8730-579586e1f497.png)

3. On the prod all lists in the trigger, add/remove tags and lists are open when you are adding those actions, on this branch they are closed, I am not sure how it is better
![image](https://user-images.githubusercontent.com/113185437/225319284-69e6e312-dfea-47d4-9c6a-986bd7187f87.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225319335-688b037e-77ae-41c9-868d-e48bd36122d8.png)

4. Text in some field names is in capslock
![image](https://user-images.githubusercontent.com/113185437/225320200-a2bd1b12-66e1-4da5-88b0-9431f580dde4.png)
![image](https://user-images.githubusercontent.com/113185437/225320252-8a8bf05b-d041-4d16-9c22-fb810c7afb2c.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225320296-340c8585-7419-465c-a4e2-dd6c158daa50.png)

5. Name of the automation is blue on this branch
![image](https://user-images.githubusercontent.com/113185437/225320399-2a49028d-5751-46d2-a260-cb2698216986.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225320450-bd8a82a1-8aae-4cae-887b-302d5de9baa3.png)

6. Issue with the + button in the form editor
![image](https://user-images.githubusercontent.com/113185437/225320524-c5aa560f-5cde-408b-88c0-5b934580108b.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225320562-8f13285a-f9b0-4a56-9fbd-eb1f4d4746bb.png)

7. No space in the size heading
![image](https://user-images.githubusercontent.com/113185437/225320651-0550b9dd-f265-4230-b88c-07b5cee5240d.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225320690-b1778392-4b8b-46b0-b284-86f61d5275e7.png)

8. Font family is bold on prod
![image](https://user-images.githubusercontent.com/113185437/225320755-8ef3a29b-7a26-48a5-b216-08250ec14214.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225320801-76f8a3c5-cebe-413f-9deb-dfb1839cf420.png)

9. Lists in the colors and message colors are different from the prod, and should be disabled when there is no chosen color
![image](https://user-images.githubusercontent.com/113185437/225320974-35d4d6d7-1a6b-409d-9f78-fdd04118dd96.png)

10. Different UI of buttons in the color settings
![image](https://user-images.githubusercontent.com/113185437/225321315-ef707b13-0d1c-4b46-a8d5-f49f2720dbb6.png)
![image](https://user-images.githubusercontent.com/113185437/225321348-478cc974-ffc3-4f2e-ae56-9b69b13d9b9b.png)
On prod
![image](https://user-images.githubusercontent.com/113185437/225321485-6cc2b539-537b-45a5-98da-758e17e4aae7.png)

11. Different UI in the template settings
![image](https://user-images.githubusercontent.com/113185437/225321576-1e1edd77-8b16-4927-86a6-917e7c5840df.png)

JanJakes commented:
Thanks, @Aschepikov! Great catches 👏

> Console error when I am pressing on add a condition

This is strange, I can't replicate it. Tried both with the premium plugin enabled and disabled. Did you have the correct premium plugin JS (caches and stuff) when you were testing that? I don't know how to reproduce it. On your screenshot, I don't see the "all of", "any of" switch, so I'd guess the premium plugin is not active, or the JS is not working. Should I check it on your website ❓

Issues:

> 1. In the automations incorrect tags list

Trying to play with the tags, but can't reproduce this. Do you have any steps to reproduce? Or should I check it on your website❓

> 2. Automations background is not blurred when an element is deleting

Indeed, it seems that [Gutenberg doesn't create the blur anymore](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default). That's fine, we'll follow whatever they do ✅

> 3. On the prod all lists in the trigger, add/remove tags and lists are open when you are adding those actions, on this branch they are closed, I am not sure how it is better

That was a bug in Gutenberg components, now it's fixed, which is desired (enabled left image, disabled right image) ✅

> 4. Text in some field names is in capslock

That seems to be the Gutenberg style for labels in the sidebar. We can see it in the post editor itself, as well as in the individual components, [such as InputControl](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-inputcontrol--with-help-text) ✅

> 5. Name of the automation is blue on this branch

Good catch, this seems to differ from Gutenberg. I need to investigate 🐞

> 6. Issue with the + button in the form editor

That looks like a bug 🐞

> 7. No space in the size heading 

Also a bug 🐞

> 8. Font family is bold on prod

That seems to be a style of [the underlying Gutenberg component](https://wordpress.github.io/gutenberg/?path=/story/components-customselectcontrol--default) ✅

> 9. Lists in the colors and message colors are different from the prod, and should be disabled when there is no chosen color

This seems to work exactly as in the WP post editor for me, so I'd say it's OK ✅

<div style="display: grid; gap: 32px;">
<img width="250" alt="Screenshot 2023-03-15 at 15 15 19" src="https://user-images.githubusercontent.com/141436/225336704-ee9e7a11-8216-4e9b-97a1-85d769050d0f.png">

<img width="250" alt="Screenshot 2023-03-15 at 15 15 28" src="https://user-images.githubusercontent.com/141436/225336718-a4587a41-e161-4c3c-a2f5-6adf25469a05.png">
</div>

> 10. Different UI of buttons in the color settings

Same here, looks like WP post editor, so I think OK ✅

> 11. Different UI in the template settings

We could fix this, but isn't the new UI actually better? It's probably more translation-friendly as well (more space). I'd keep it on two lines ✅

Aschepikov commented:
@JanJakes  For the first error you can check it here https://alex.qawp.net/wp-admin/admin.php?page=mailpoet-segments#/new-segment
I reinstalled plugins and cleaned all cache, I mentioned in slack that it is reproducible just on this site, maybe some problem with site

Aschepikov commented:
With tags, I can't reproduce it after reinstalling too, maybe there was a lag with installing or a cache problem. 
I agree with all your comments. 

Just the 4 with capslock, there are some headings without capslock and I remember that there was a ticket with fixing some capslock headings in forms editors. Looks a lit bit strange for me this capslock, but yes I checked it in wp posts and they have it too.
![image](https://user-images.githubusercontent.com/113185437/225350656-84d939f2-c65b-42b7-877c-ec1d5fab1265.png)
![image](https://user-images.githubusercontent.com/113185437/225350768-2aa3d328-cda0-4cc6-8e21-6efde20df7fb.png)

JanJakes committed:
Fix form editor block inserter styles after Gutenberg upgrade

[MAILPOET-5102]

JanJakes committed:
Fix form editor style settings styles after Gutenberg upgrade

[MAILPOET-5102]

JanJakes committed:
Fix automation name button style

[MAILPOET-5102]

JanJakes committed:
Fix form editor global styles font size picker

[MAILPOET-5102]

JanJakes commented:
@Aschepikov I fixed 5, 6, 7, and also 11, and another small issue.

> Just the 4 with capslock, there are some headings without capslock and I remember that there was a ticket with fixing some capslock headings in forms editors. Looks a lit bit strange for me this capslock, but yes I checked it in wp posts and they have it too.

Headers in the panels are not uppercased. It's only labels that are uppercased.

Just need to check the first issue and all else should be fine.

JanJakes commented:
@Aschepikov I also likely fixed number 1 — improved the code to avoid any race conditions or duplicates. I also found a new issue (that the first time the "add tag" select was focused, it showed no autocomplete) and fixed it too.

JanJakes committed:
Fix radio control option alignment

[MAILPOET-5102]

Aschepikov reviewed (approved):
Great job!
